tests/utils/engine_mocking.py

```diff
@@ -0,0 +1,181 @@
+    with `_FakeDeepsparseLibEngine`.
+
+    Use like you would a regular `unittest.mock.patch`
+    ```python
+    @mock_engine(rng_seed=0)
+    def test_something(engine_mock: MagicMock):
+        ...
+    ```
+
+    If you want the arrays to be zeros instead of random data:
+    ```python
+    @mock_engine(rng_seed=0, model=SampleMode.zeros)
+    def test_something(engine_mock: MagicMock):
+        ...
+    ```
+
+    The object that is passed into unit tests is a
+    [unittest.MagicMock](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.MagicMock)
+    object.
+
+    You can use it to see how many times LIB.deepsparse_engine() was invoked and
+    with what arguments. For example of how to use this, see the
+    `test_engine_mocking.py` file.
+    """
+
+    def deepsparse_engine(*args, **kwargs):
+        return _FakeDeepsparseLibEngine(*args, **kwargs, rng_seed=rng_seed, mode=mode)
+
+    return mock.patch(
+        "deepsparse.engine.LIB.deepsparse_engine",
+        side_effect=deepsparse_engine,
+        **kwargs,
+    )
+
+
+class _FakeDeepsparseLibEngine:
+    def __init__(
+        self,
+        model_path: str,
+        batch_size: int,
+        num_cores: int,
+        num_streams: int,
+        scheduler_value: Any,
+        context_value: Any,
+        *,
+        # NOTE: the following are not actual deepsparse engine arguments
+        rng_seed: int,
+        mode: SampleMode,
+    ):
+        self.model_path = model_path
+        self.batch_size = batch_size
+        self.num_cores = num_cores
+        self.num_streams = num_streams
+        self.scheduler_value = scheduler_value
+        self.context_value = context_value
+        self.rng_seed = rng_seed
+        self.mode = mode
+
+        # override batch dimension in inputs & outputs
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            model = onnx.load(self.model_path)
+            inputs = model.graph.input
+            initializer_names = [node.name for node in model.graph.initializer]
+            # remove initializer nodes from inputs
+            inputs = [input for input in inputs if input.name not in initializer_names]
+
+            # This is done because not all `.onnx` files passed in will
+            # conform to expected `batch_size`.
+            #
+            # By overriding the batch size in the inputs, and loading this
+            # model with the onnxruntime, onnxruntime will automatically
+            # fix the batch dimension in outputs.
+            #
+            # Assumes the first dimension is batch dimension!!
+            # However in general we cannot assume that all outputs have
+            # a batch dimension, that's why we need onnxruntime here.
+            for input in inputs:
+                input.type.tensor_type.shape.dim[0].dim_value = batch_size
+
+            batched_model_path = str(Path(tmp_dir) / "batched_model.onnx")
+            onnx.save(model, batched_model_path)
```

Comment:
is there a reason to not use `override_onnx_batch_size` here? It seems to do a similar action. Understood why you need to put it through ORT later. Note that this will not give you valid shapes if some input/output dimensions are not static or the model's output shape is based on input (like variable num boxes in object detection models)

Reply:
Yes its the same functionality, but I wanted more control over the path that override_onnx_batch_size saved to (so it was under `tmp_dir`). Probably not a good enough reason to duplicate here in retrospect, I'll use the existing function!